cupy/cuda/cupy_thrust.cu

```diff
@@ -43,6 +44,102 @@ public:
 };
 
 
+/* ------------------------------------ Minimum boilerplate to support complex numbers ------------------------------------ */
+// We need a specialized operator< here in order to match the NumPy behavior:
+// "The sort order for complex numbers is lexicographic. If both the real and imaginary parts are non-nan then the order is
+// determined by the real parts except when they are equal, in which case the order is determined by the imaginary parts.
+//
+// In numpy versions >= 1.4.0 nan values are sorted to the end. The extended sort order is:
+//     Real: [R, nan]
+//     Complex: [R + Rj, R + nanj, nan + Rj, nan + nanj]
+// where R is a non-nan real value. Complex values with the same nan placements are sorted according to the non-nan part if
+// it exists. Non-nan values are sorted as before."
+// Ref: https://docs.scipy.org/doc/numpy/reference/generated/numpy.sort.html
+
+__host__ __device__ inline bool isnan(const cuComplex& z) {
+    return isnan(z.x) || isnan(z.y);
+}
+
+__host__ __device__ inline bool isnan(const cuDoubleComplex& z) {
+    return isnan(z.x) || isnan(z.y);
+}
+
+__host__ __device__ inline bool operator<(const cuComplex& lhs,
+                                          const cuComplex& rhs) {
+  if (isnan(lhs)) {
+      if (!isnan(rhs)) {
+          return false;
+      } else if (isnan(lhs.x) && !isnan(rhs.x)) {
+          return false;
+      } else if (!isnan(lhs.x) && isnan(rhs.x)) {
+          return true;
+      } else if (isnan(lhs.y) && !isnan(rhs.y)) {
+          return false;
+      } else if (!isnan(lhs.y) && isnan(rhs.y)) {
+          return true;
+      } else if (isnan(lhs.y) && isnan(rhs.y)) {
+          return lhs.x < rhs.x;
+      } else if (isnan(lhs.x) && isnan(rhs.x)) {
+          return lhs.y < rhs.y;
+      } else { // both lhs & rhs = nan + nan I
+          return true;
```

Comment:
Instead of crazy branching, how about a single boolean expression?

Reply:
What do you mean?

Reply:
‘’’python
bool less_than = is_nan(lhs.x) && .....
return less_than;
‘’’

Reply:
I see, thanks. Let me try.

Reply:
I was not able to come up with a cleaner comparison logic to handle the NumPy rule. In 668efd0 I made the NaN handling templated. Any suggestion, @emcastillo?

Reply:
No worries, the actual code can deal with this fine. Let's revisit this at another time.